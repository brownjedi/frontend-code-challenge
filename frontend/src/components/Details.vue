<template>
  <div :class="styles.container">
    <transition name="fade">
      <!-- Loading -->
      <div v-if="loading" :class="styles.loading">
        <img src="/img/loading.gif" alt="loading" />
      </div>

      <!-- Error -->
      <div v-else-if="error" :class="styles.error">An error occured</div>

      <!-- Result -->
      <div v-else-if="pokemon" :class="styles.result" :key="pokemon.id">
        <div :class="styles.left">
          <div :class="[styles.img, styles.innerContainer]">
            <img :src="pokemon.image" :alt="pokemon.name" @click="playAudio" />
          </div>
        </div>
        <div :class="styles.right">
          <div :class="[styles.info, styles.innerContainer]">
            <div :class="styles.idNameAudioContainer">
              <div :class="styles.nameAudioContainer">
                <div :class="styles.name">{{ pokemon.name }}</div>
                <div :class="styles.audio">
                  <img src="/img/document--audio.svg" alt="pokemon audio" @click="playAudio" />
                </div>
              </div>
              <div :class="styles.id">#{{ pokemon.id }}</div>
            </div>
            <div :class="styles.classification">{{ pokemon.classification }}</div>
            <p>{{ pokemon.isFavorite }}</p>
            <p>{{ pokemon.maxCP }}</p>
            <p>{{ pokemon.maxHP }}</p>
            <p>{{ pokemon.fleeRate }}</p>
            <div :class="styles.tagsContainer">
              <div :class="styles.label">Resistances</div>
              <Tags :tags="pokemon.resistant" :class="styles.tags" />
            </div>
            <div :class="styles.tagsContainer">
              <div :class="styles.label">Types</div>
              <Tags :tags="pokemon.types" :class="styles.tags" />
            </div>
            <div :class="styles.tagsContainer">
              <div :class="styles.label">Weaknesses</div>

              <Tags :tags="pokemon.weaknesses" :class="styles.tags" />
            </div>
            <Favorite :class="styles.favorite" :large="true" :isFavorite="pokemon.isFavorite" @click="toggleFavorite" />
            <audio ref="audioRef">
              <source :src="pokemon.sound" type="audio/mp3" />
              Your browser does not support the audio element.
            </audio>
          </div>
        </div>
      </div>
      <!-- No result -->
      <div v-else :class="styles.noResult">No result :(</div>
    </transition>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, ref, useCssModule } from 'vue'
import Tags from './Tags.vue'
import Favorite from './Favorite.vue'
import { usePokemonByNameQuery } from '@/composables/pokemonByNameQuery'
import { FAVORITE_POKEMON_MUTATION, UNFAVORITE_POKEMON_MUTATION } from '@/graphql/graphql-queries'
import { useMutation } from '@vue/apollo-composable'

export default defineComponent({
  components: {
    Tags,
    Favorite,
  },
  props: {
    name: {
      type: String,
      required: true,
    },
  },
  setup(props) {
    const styles = useCssModule()
    const audioRef = ref(null)
    const name = computed(() => props.name.toLowerCase())
    const { pokemon, loading, error } = usePokemonByNameQuery(name.value)
    const { mutate: favoritePokemon } = useMutation(FAVORITE_POKEMON_MUTATION)
    const { mutate: unFavoritePokemon } = useMutation(UNFAVORITE_POKEMON_MUTATION)

    function playAudio() {
      const el = audioRef.value as unknown as HTMLAudioElement
      el.play()
    }

    function toggleFavorite() {
      if (pokemon.value) {
        pokemon.value.isFavorite
          ? unFavoritePokemon({ id: pokemon.value.id })
          : favoritePokemon({ id: pokemon.value.id })
      }
    }

    const backgroundColor = computed(() => pokemon.value?.rgbaster?.backgroundColor || 'inherit')
    const textColor = computed(() => pokemon.value?.rgbaster?.textColor || 'inherit')

    return { styles, pokemon, loading, error, backgroundColor, textColor, audioRef, playAudio, toggleFavorite }
  },
})
</script>

<style lang="scss" module>
.container {
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
}

.result {
  flex: 1;
  display: flex;
  align-self: stretch;

  .inner-container {
    max-width: 80rem;
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    margin: 3rem;
  }
}

.left {
  transition: background-color 0.5s ease-in-out, color 0.5s ease-in-out;
  color: v-bind(textColor);
  background-color: v-bind(backgroundColor);
  flex: 1 1 50%;
  display: flex;
  justify-content: flex-end;

  img {
    display: block;
    width: 100%;
    height: auto;
    max-width: 40rem;
    cursor: pointer;
  }
}

.right {
  flex: 1 1 50%;
  display: flex;
  justify-content: flex-start;

  .info {
    flex-direction: column;
    justify-content: center;
  }

  .tags-container {
    align-self: flex-start;
    padding-top: 2rem;

    .tags {
      position: relative;
      padding-top: 1rem;
    }

    .label {
      font-weight: 500;
    }
  }

  .id-name-audio-container {
    align-self: stretch;
    display: flex;
    justify-content: space-between;
    font-size: 3rem;
  }

  .name-audio-container {
    display: flex;
  }

  .id {
    align-self: flex-end;
    color: var(--number-title-color);
    font-weight: 400;
  }

  .name {
    font-weight: 400;
    margin-right: 0.25rem;
  }

  .audio {
    cursor: pointer;
    padding-top: 0.125rem;
    img {
      width: 2rem;
      height: 2rem;
    }
  }

  .classification {
    padding-top: 0.25rem;
    align-self: flex-start;
    font-weight: 300;
  }

  .favorite {
    margin-top: 3rem;
    align-self: center;
  }
}

.error {
  color: red;
}
</style>
