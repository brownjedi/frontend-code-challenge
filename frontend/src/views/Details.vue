<template>
  <div>
    <PokemonDetails />
    <button @click="previous">Previous</button>
    <button @click="next">Next</button>
  </div>
</template>

<script>
import PokemonDetails from '@/components/PokemonDetails.vue'
export default {
  components: {
    PokemonDetails,
  },
  methods: {
    previous() {
      console.log('previous')
    },
    next() {
      console.log('next')
    },
  },
}
</script>

<style lang="scss"></style>
