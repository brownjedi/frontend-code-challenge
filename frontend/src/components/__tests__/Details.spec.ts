import { flushPromises, shallowMount } from '@vue/test-utils'
import Details from '@/components/Details.vue'
import { createMockClient, MockApolloClient } from 'mock-apollo-client'
import {
  FAVORITE_POKEMON_MUTATION,
  POKEMON_BY_NAME_QUERY,
  UNFAVORITE_POKEMON_MUTATION,
} from '@/graphql/graphql-queries'
import { provideApolloClient } from '@vue/apollo-composable'

const pokemon = {
  id: '001',
  name: 'Bulbasaur',
  image: 'http://localhost:4000/img/artwork/vector/large/bulbasaur.png',
  sprites: {
    normal: 'http://localhost:4000/img/sprites/black-white/normal/bulbasaur.png',
    animated: 'http://localhost:4000/img/sprites/black-white/anim/normal/bulbasaur.gif',
  },
  classification: 'Seed PokÃ©mon',
  types: ['Grass', 'Poison'],
  isFavorite: false,
  number: 1,
  maxCP: 951,
  maxHP: 1071,
  fleeRate: 0.1,
  resistant: ['Water', 'Electric', 'Grass', 'Fighting', 'Fairy'],
  sound: 'http://localhost:4000/sounds/1',
  weaknesses: ['Fire', 'Ice', 'Flying', 'Psychic'],
  attacks: {
    special: [
      {
        name: 'Power Whip',
        type: 'Grass',
        damage: 70,
      },
      {
        name: 'Seed Bomb',
        type: 'Grass',
        damage: 40,
      },
      {
        name: 'Sludge Bomb',
        type: 'Poison',
        damage: 55,
      },
    ],
    fast: [
      {
        name: 'Tackle',
        type: 'Normal',
        damage: 12,
      },
      {
        name: 'Vine Whip',
        type: 'Grass',
        damage: 7,
      },
    ],
  },
  evolutionRequirements: {
    name: 'Bulbasaur candies',
    amount: 25,
  },
  evolutions: [
    {
      id: '002',
      name: 'Ivysaur',
    },
    {
      id: '003',
      name: 'Venusaur',
    },
  ],
  height: {
    minimum: '0.61m',
    maximum: '0.79m',
  },
  weight: {
    minimum: '6.04kg',
    maximum: '7.76kg',
  },
}

describe('components/Details.vue', () => {
  let favRequestHandler: jest.Mock
  let unFavRequestHandler: jest.Mock
  let pokemonByNameRequestHandler: jest.Mock
  let mockClient: MockApolloClient

  beforeEach(() => {
    jest.resetAllMocks()
  })

  beforeEach(() => {
    mockClient = createMockClient()
    pokemonByNameRequestHandler = jest.fn()
    favRequestHandler = jest.fn()
    unFavRequestHandler = jest.fn()
    mockClient.setRequestHandler(POKEMON_BY_NAME_QUERY, pokemonByNameRequestHandler)
    mockClient.setRequestHandler(FAVORITE_POKEMON_MUTATION, favRequestHandler)
    mockClient.setRequestHandler(UNFAVORITE_POKEMON_MUTATION, unFavRequestHandler)
    provideApolloClient(mockClient)
  })

  describe('when query loading', () => {
    it('has the expected html structure', () => {
      const wrapper = shallowMount(Details, {
        props: {
          name: pokemon.name,
        },
      })
      expect(wrapper.element).toMatchSnapshot()
    })

    it('sends the pokemon name in the graphql query', () => {
      shallowMount(Details, {
        props: {
          name: pokemon.name,
        },
      })

      expect(pokemonByNameRequestHandler).toHaveBeenCalledWith({ name: pokemon.name.toLowerCase() })
    })
  })

  describe('when query fails', () => {
    it('shows error message', async () => {
      pokemonByNameRequestHandler.mockRejectedValueOnce(new Error('Graphql query failed'))

      const wrapper = shallowMount(Details, {
        props: {
          name: pokemon.name,
        },
      })

      expect(pokemonByNameRequestHandler).toHaveBeenCalledTimes(1)

      // wait for apollo results
      await flushPromises()

      // ensure that result container exists
      expect(wrapper.find('.error').exists()).toBe(true)
      expect(wrapper.find('.error').text()).toBe('An error occured')
    })
  })

  describe('when query succeeds', () => {
    it('shows no result message when no data', async () => {
      pokemonByNameRequestHandler.mockResolvedValueOnce({
        data: { pokemon: null },
      })

      const wrapper = shallowMount(Details, {
        props: {
          name: pokemon.name,
        },
      })

      expect(pokemonByNameRequestHandler).toHaveBeenCalledTimes(1)

      // wait for apollo results
      await flushPromises()

      // ensure that result container exists
      expect(wrapper.find('.no-result').exists()).toBe(true)
      expect(wrapper.find('.no-result').text()).toBe('No result :(')
    })

    it('favorite the pokemon', async () => {
      pokemonByNameRequestHandler.mockResolvedValueOnce({
        data: { pokemon },
      })
      favRequestHandler.mockResolvedValue({ data: { pokemon: { id: pokemon.id, isFavorite: true } } })
      unFavRequestHandler.mockResolvedValue({ data: { pokemon: { id: pokemon.id, isFavorite: false } } })

      const wrapper = shallowMount(Details, {
        props: {
          name: pokemon.name,
        },
      })

      // wait for apollo results
      await flushPromises()

      // trigger to favorite it
      await wrapper.find('favorite-stub').trigger('click')

      expect(favRequestHandler).toHaveBeenCalledTimes(1)
      expect(unFavRequestHandler).toHaveBeenCalledTimes(0)
    })

    it('unfavorite the pokemon', async () => {
      pokemonByNameRequestHandler.mockResolvedValueOnce({
        data: { pokemon: { ...pokemon, isFavorite: true } },
      })
      favRequestHandler.mockResolvedValue({ data: { pokemon: { id: pokemon.id, isFavorite: true } } })
      unFavRequestHandler.mockResolvedValue({ data: { pokemon: { id: pokemon.id, isFavorite: false } } })

      const wrapper = shallowMount(Details, {
        props: {
          name: pokemon.name,
        },
      })

      // wait for apollo results
      await flushPromises()

      // trigger to favorite it
      await wrapper.find('favorite-stub').trigger('click')

      expect(favRequestHandler).toHaveBeenCalledTimes(0)
      expect(unFavRequestHandler).toHaveBeenCalledTimes(1)
    })

    it('renders results correctly when data is present', async () => {
      pokemonByNameRequestHandler.mockResolvedValueOnce({
        data: { pokemon },
      })

      const wrapper = shallowMount(Details, {
        props: {
          name: pokemon.name,
        },
      })

      // wait for apollo results
      await flushPromises()

      expect(wrapper.element).toMatchSnapshot()
    })

    it('plays audio when clicked on the img', async () => {
      const playStub = jest.spyOn(window.HTMLMediaElement.prototype, 'play').mockResolvedValue()
      pokemonByNameRequestHandler.mockResolvedValueOnce({
        data: { pokemon },
      })

      const wrapper = shallowMount(Details, {
        props: {
          name: pokemon.name,
        },
      })

      // wait for apollo results
      await flushPromises()

      const imgEl = wrapper.findAll('img')

      // there are only two images once the data is loaded (for now)
      // one is the pokemon image and other is the speaker icon
      // both trigger audio
      expect(imgEl.length).toBe(2)

      for (let i = 0; i < imgEl.length; i++) {
        await imgEl[i].trigger('click')
        expect(playStub).toHaveBeenCalled()
      }

      expect(playStub).toHaveBeenCalledTimes(2)
      playStub.mockRestore()
    })
  })
})
