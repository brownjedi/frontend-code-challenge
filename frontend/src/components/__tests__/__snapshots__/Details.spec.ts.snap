// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Details.vue when query loading has the expected html structure 1`] = `
<div
  class="container"
>
  <transition-stub>
    <div
      class="loading"
    >
      <img
        alt="loading"
        src="/img/loading.gif"
      />
    </div>
  </transition-stub>
</div>
`;

exports[`components/Details.vue when query succeeds renders results correctly when data is present 1`] = `
<div
  class="container"
>
  <transition-stub>
    <div
      class="result"
    >
      <div
        class="left"
      >
        <div
          class="inner-container"
        >
          <img
            alt="Bulbasaur"
            data-cy="img-play-audio"
            src="http://localhost:4000/img/artwork/vector/large/bulbasaur.png"
          />
        </div>
      </div>
      <div
        class="right"
      >
        <div
          class="info inner-container"
        >
          <div
            class="id-name-audio-container"
          >
            <div
              class="name-audio-container"
            >
              <div
                class="name"
              >
                Bulbasaur
              </div>
              <div
                class="audio"
              >
                <img
                  alt="pokemon audio"
                  data-cy="img-play-audio"
                  src="/img/document--audio.svg"
                />
              </div>
            </div>
            <div
              class="id"
            >
              #001
            </div>
          </div>
          <div
            class="classification"
          >
            Seed Pok√©mon
          </div>
          <div
            class="stats"
          >
            <div
              class="stat"
            >
              <div
                class="label"
              >
                Max CP
              </div>
              <p>
                951
              </p>
            </div>
            <div
              class="stat"
            >
              <div
                class="label"
              >
                Max HP
              </div>
              <p>
                1071
              </p>
            </div>
            <div
              class="stat"
            >
              <div
                class="label"
              >
                Flee Rate
              </div>
              <p>
                0.1
              </p>
            </div>
            <div
              class="stat"
            >
              <div
                class="label"
              >
                Height
              </div>
              <p>
                0.61m - 0.79m
              </p>
            </div>
            <div
              class="stat"
            >
              <div
                class="label"
              >
                Weight
              </div>
              <p>
                6.04kg - 7.76kg
              </p>
            </div>
          </div>
          <hr
            class="hr"
          />
          <div
            class="attacks"
          >
            <div
              class=""
            >
              <div
                class="label"
              >
                Special Attacks
              </div>
              
              <div
                class="attack"
              >
                <img
                  alt="Grass"
                  src="/img/types/grass.png"
                />
                <p>
                  Power Whip
                </p>
              </div>
              <div
                class="attack"
              >
                <img
                  alt="Grass"
                  src="/img/types/grass.png"
                />
                <p>
                  Seed Bomb
                </p>
              </div>
              <div
                class="attack"
              >
                <img
                  alt="Poison"
                  src="/img/types/poison.png"
                />
                <p>
                  Sludge Bomb
                </p>
              </div>
              
            </div>
            <div
              class=""
            >
              <div
                class="label"
              >
                Fast Attacks
              </div>
              
              <div
                class="attack"
              >
                <img
                  alt="Normal"
                  src="/img/types/normal.png"
                />
                <p>
                  Tackle
                </p>
              </div>
              <div
                class="attack"
              >
                <img
                  alt="Grass"
                  src="/img/types/grass.png"
                />
                <p>
                  Vine Whip
                </p>
              </div>
              
            </div>
          </div>
          <hr
            class="hr"
          />
          <div
            class="tags-container"
          >
            <div
              class="label"
            >
              Types
            </div>
            <tags-stub
              class="tags"
              tags="Grass,Poison"
            />
          </div>
          <div
            class="tags-container"
          >
            <div
              class="label"
            >
              Resistances
            </div>
            <tags-stub
              class="tags"
              tags="Water,Electric,Grass,Fighting,Fairy"
            />
          </div>
          <div
            class="tags-container"
          >
            <div
              class="label"
            >
              Weaknesses
            </div>
            <tags-stub
              class="tags"
              tags="Fire,Ice,Flying,Psychic"
            />
          </div>
          <hr
            class="hr"
          />
          <div
            class="evolutions"
          >
            <div
              class="label"
            >
              Evolutions
            </div>
            <evolutions-stub
              evolutions="[object Object],[object Object],[object Object]"
            />
          </div>
          <hr
            class="hr"
          />
          <favorite-stub
            class="favorite"
            isfavorite="false"
            large="true"
          />
          <audio>
            <source
              src="http://localhost:4000/sounds/1"
              type="audio/mp3"
            />
             Your browser does not support the audio element. 
          </audio>
        </div>
      </div>
    </div>
  </transition-stub>
</div>
`;
